File: src/main/java/at/yawk/wm/tac/password/TextEditorWindow.java
package at.yawk.wm.tac.password;

import at.yawk.wm.style.FontManager;
import at.yawk.wm.style.FontStyle;
import java.awt.*;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import javax.swing.*;
import lombok.Getter;

/**
 * @author yawkat
 */
@Getter
class TextEditorWindow {
    private final Window window;
    private final JTextArea field;

    public TextEditorWindow(FontManager fontManager, PasswordConfig config, String text) {
        Dimension size = new Dimension(config.getEditorWidth(), config.getEditorHeight());

        field = new JTextArea() {
            @Override
            protected void paintComponent(Graphics g) {
                if (g instanceof Graphics2D) {
                    ((Graphics2D) g).setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING,
                                                      RenderingHints.VALUE_TEXT_ANTIALIAS_LCD_HBGR);
                }
                super.paintComponent(g);
            }
        };
        field.setText(text);
        field.setBackground(config.getEditorBackground());
        FontStyle style = fontManager.resolve(config.getEditorFont());
        field.setFont(style.getFamily().createFont(style));
        field.setForeground(style.getForeground());
        field.setSelectedTextColor(config.getEditorBackground());
        field.setSelectionColor(style.getForeground());
        field.setCaretColor(style.getForeground());
        field.setLineWrap(true);

        window = new JDialog();
        window.setSize(size);
        window.setBackground(config.getEditorBackground());
        window.setLayout(new BorderLayout());
        window.add(field, BorderLayout.CENTER);
        window.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                close();
            }
        });
    }

    public String getText() {
        return field.getText();
    }

    public void show() {
        window.setVisible(true);
    }

    public void close() {
        field.setText(""); // clear text
        window.dispose();
    }

    public void setTitle(String title) {
        if (window instanceof Frame) {
            ((Frame) window).setTitle(title);
        } else if (window instanceof Dialog) {
            ((Dialog) window).setTitle(title);
        }
    }
}

File: src/main/java/at/yawk/wm/tac/password/TextEditorWindow.kt
package at.yawk.wm.tac.password

import at.yawk.wm.x.font.AwtGlyphFileFactory
import java.awt.BorderLayout
import java.awt.Dialog
import java.awt.Dimension
import java.awt.Frame
import java.awt.Window
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import javax.swing.JDialog
import javax.swing.JTextArea
import javax.swing.SwingUtilities
import javax.swing.UIManager

internal open class TextEditorWindow {
    val window: Window
    private val field: JTextArea

    val text: String
        get() = this.field.text

    constructor(text: String) {
        this.field.text = text
    }

    fun show() {
        SwingUtilities.invokeLater { window.isVisible = true }
    }

    open fun close() {
        SwingUtilities.invokeLater {
            field.text = ""
            window.dispose()
        }
    }

    fun setTitle(title: String?) {
        if (window is Frame) {
            window.title = title
        } else if (window is Dialog) {
            window.title = title
        }
    }

    fun getField(): JTextArea {
        return field
    }

    init {
        initSwing()
        val size = Dimension(PasswordConfig.editorWidth, PasswordConfig.editorHeight)
        val style = PasswordConfig.editorFont
        val font = AwtGlyphFileFactory.getFont(style)
        field = JTextArea()
        field.setText(text)
        field.setBackground(PasswordConfig.editorBackground.awt)
        field.setFont(font)
        field.setForeground(style.foreground.awt)
        field.setSelectedTextColor(PasswordConfig.editorBackground.awt)
        field.setSelectionColor(style.foreground.awt)
        field.setCaretColor(style.foreground.awt)
        field.setLineWrap(true)
        window = JDialog()
        window.size = size
        window.setBackground(PasswordConfig.editorBackground.awt)
        window.layout = BorderLayout()
        window.add(field, BorderLayout.CENTER)
        window.addWindowListener(object : WindowAdapter() {
            override fun windowClosing(e: WindowEvent) {
                close()
            }
        })
        window.isLocationByPlatform = true
    }

    companion object {
        private val log = org.slf4j.LoggerFactory.getLogger(TextEditorWindow::class.java)
        private var swingInitialized = false

        @Synchronized
        private fun initSwing() {
            if (!swingInitialized) {
                swingInitialized = true
                log.info("Initializing Swing: DISPLAY={}", System.getenv("DISPLAY"))
                try {
                    UIManager.setLookAndFeel(UIManager.getCrossPlatformLookAndFeelClassName())
                    log.info("Set cross-platform LookAndFeel successfully")
                } catch (e: Exception) {
                    log.warn("Failed to set cross-platform LookAndFeel", e)
                }
            }
        }
    }
}
